v2/yarpchttp/handler.go

```diff
@@ -52,193 +49,197 @@ type handler struct {
+		responseWriter.SetResponse(res, yarpc.NewBufferString(status.Message()))
 	}
+
 	httpStatusCode, ok := _codeToStatusCode[status.Code()]
 	if !ok {
 		httpStatusCode = http.StatusInternalServerError
 	}
 	responseWriter.Close(httpStatusCode)
 }
 
-func (h handler) callHandler(responseWriter *responseWriter, req *http.Request, service string, procedure string) (retErr error) {
+func (h handler) callHandler(
+	responseWriter *responseWriter,
+	httpReq *http.Request,
+	service string,
+	procedure string,
+) (res *yarpc.Response, resBuf *yarpc.Buffer, err error) {
 	start := time.Now()
-	defer req.Body.Close()
-	if req.Method != http.MethodPost {
-		return yarpcerror.Newf(yarpcerror.CodeNotFound, "request method was %s but only %s is allowed", req.Method, http.MethodPost)
+	if httpReq.Method != http.MethodPost {
+		return nil, nil, yarpcerror.Newf(yarpcerror.CodeNotFound, "request method was %s but only %s is allowed", httpReq.Method, http.MethodPost)
 	}
-	treq := &yarpc.Request{
-		Caller:          popHeader(req.Header, CallerHeader),
+	req := &yarpc.Request{
+		Caller:          popHeader(httpReq.Header, CallerHeader),
 		Service:         service,
 		Procedure:       procedure,
-		Encoding:        yarpc.Encoding(popHeader(req.Header, EncodingHeader)),
+		Encoding:        yarpc.Encoding(popHeader(httpReq.Header, EncodingHeader)),
 		Transport:       transportName,
-		ShardKey:        popHeader(req.Header, ShardKeyHeader),
-		RoutingKey:      popHeader(req.Header, RoutingKeyHeader),
-		RoutingDelegate: popHeader(req.Header, RoutingDelegateHeader),
-		Headers:         applicationHeaders.FromHTTPHeaders(req.Header, yarpc.Headers{}),
-		Body:            req.Body,
+		ShardKey:        popHeader(httpReq.Header, ShardKeyHeader),
+		RoutingKey:      popHeader(httpReq.Header, RoutingKeyHeader),
+		RoutingDelegate: popHeader(httpReq.Header, RoutingDelegateHeader),
+		Headers:         applicationHeaders.FromHTTPHeaders(httpReq.Header, yarpc.Headers{}),
 	}
+
+	reqBuf, err := readCloserToBuffer(httpReq.Body)
+	if err != nil {
+		return nil, nil, err
+	}
+
 	for header := range h.grabHeaders {
-		if value := req.Header.Get(header); value != "" {
-			treq.Headers = treq.Headers.With(header, value)
+		if value := httpReq.Header.Get(header); value != "" {
+			req.Headers = req.Headers.With(header, value)
 		}
 	}
-	if err := yarpc.ValidateRequest(treq); err != nil {
-		return err
+	if err = yarpc.ValidateRequest(req); err != nil {
+		return nil, nil, err
 	}
-	defer func() {
-		if retErr == nil {
-			if contentType := getContentType(treq.Encoding); contentType != "" {
-				responseWriter.AddSystemHeader("Content-Type", contentType)
-			}
-		}
-	}()
 
-	ctx := req.Context()
-	ctx, cancel, parseTTLErr := parseTTL(ctx, treq, popHeader(req.Header, TTLMSHeader))
+	ctx := httpReq.Context()
+	ctx, cancel, parseTTLErr := parseTTL(ctx, req, popHeader(httpReq.Header, TTLMSHeader))
 	// parseTTLErr != nil is a problem only if the request is unary.
```

Comment:
Is this comment out-of-date? It looks like we check the `if parseTTLErr != nil` conditional below.